UI/window-basic-main.cpp

```diff
@@ -1603,6 +1621,314 @@ void OBSBasic::UpdateMultiviewProjectorMenu()
+
+err_av_image_alloc:
+	// Failed allocating image data buffer
+	av_frame_free(&frame);
+	frame = NULL;
+
+err_av_frame_alloc:
+	// Failed allocating frame
+	avcodec_close(codec_context);
+
+err_png_encoder_open:
+	// Failed opening PNG encoder
+	avcodec_free_context(&codec_context);
+	codec_context = NULL;
+
+err_png_encoder_context_alloc:
+	// failed allocating PNG encoder context
+	// no need to free AVCodec* codec
+err_png_codec_not_found:
+	// PNG encoder not found
+err_no_image_data:
+	// image_data_ptr == NULL
+
+	return success;
+}
+
+bool OBSBasic::SaveScreenshotToFile(const char* outputFilePath)
+{
+	obs_video_info video_info;
+
+	if (!obs_get_video_info(&video_info))
+		return false;
+
+	// Setup raw video format settings
+	video_scale_info raw_video_format;
+
+	raw_video_format.width = video_info.base_width;
+	raw_video_format.height = video_info.base_height;
+	raw_video_format.format = VIDEO_FORMAT_RGBA;
+	raw_video_format.range = VIDEO_RANGE_DEFAULT;
+
+	// This structure holds data interchanged with the
+	// raw video callback function
+	struct raw_video_callback_frame_data_args {
+		uint32_t   width; // frame width
+		uint32_t   height; // frame height
+		uint8_t    *image_data_ptr; // actual image data
+		size_t     image_data_buffer_length; // image data size
+		uint32_t   image_data_linesize; // image data pitch
+		os_event_t *complete_event; // event fired when done
+
+		// reference to the callback function
+		void(*raw_video_callback_ref)(void *, struct video_data *);
+	};
+
+	raw_video_callback_frame_data_args callback_data;
+
+	// initialize completion event
+	os_event_init(&callback_data.complete_event, OS_EVENT_TYPE_AUTO);
+
+	// initialize callback function input
+	callback_data.width = video_info.base_width;
+	callback_data.height = video_info.base_height;
+
+	// set up our callback function body for obs_add_raw_video_callback()
+	callback_data.raw_video_callback_ref =
+		[](void *param, video_data* data)
+	{
+		raw_video_callback_frame_data_args *args =
+			(raw_video_callback_frame_data_args *)param;
+
+		// Copy screenshot data aside
+		args->image_data_linesize = data->linesize[0];
+		args->image_data_buffer_length =
+			args->image_data_linesize * args->height;
+
+		// Allocate memory for video data
+		args->image_data_ptr =
+			(uint8_t*)bmalloc(
+				args->image_data_buffer_length);
+
+		// Copy video data to our own buffer
+		memcpy(
+			args->image_data_ptr, // destination buffer
+			data->data[0], // source buffer
+			args->image_data_buffer_length);
+
+		// Remove our callback as soon as we're done
+		obs_remove_raw_video_callback(
+			args->raw_video_callback_ref,
+			param);
+
+		// Signal the calling thread we're done
+		os_event_signal(args->complete_event);
+	};
+
+	// Add raw video callback to get the next video frame
+	obs_add_raw_video_callback(
+		&raw_video_format, // requested video format
+		callback_data.raw_video_callback_ref, // callback
+		&callback_data // callback arguments
+	); 
```

Comment:
Trailing whitespace on this line.  Also, Inconsistent code styling with the rest of the project when calling this function. Should be:

```C
	obs_add_raw_video_callback(
			&raw_video_format,
			callback_data.raw_video_callback_ref,
			&callback_data);
```